MINOR: Send latest LeaderAndIsr version

stanislavkozlovski opened the pull request:
KIP-455 (18d4e57f6e8c67ffa7937fc855707d3a03cc165a) bumped the LeaderAndIsr version to 3 but did not change the Controller code to actually send the new version. The ControllerChannelManagerTest had a bug which made it assert wrongly, hence why it did not catch it. This patch fixes said test.
Because the new fields in LeaderAndIsr are not used yet, the gap was not caught by integration tests either.

stanislavkozlovski committed:
MINOR: Send latest LeaderAndIsr version
KIP-455 (18d4e57f6e8c67ffa7937fc855707d3a03cc165a) bumped the LeaderAndIsr version to 3 but did not change the Controller code to actually send the new version. Because the new fields in LeaderAndIsr are not used yet, the gap was not caught by tests.
Due to the backwards and forwards compatibility of the change, we do not guard the request version against inter.broker.version.protocol

stanislavkozlovski commented:
cc @cmccabe @hachikuji @junrao 

I am a bit confused regarding the `ControllerChannelManagerTest#testLeaderAndIsrRequestFollowsInterBrokerProtocolVersion()` test.

I see it builds the request like this:
```
val leaderAndIsrRequests = batch.collectLeaderAndIsrRequestsFor(2, expectedLeaderAndIsrVersion)
assertEquals(1, leaderAndIsrRequests.size)
assertEquals(expectedLeaderAndIsrVersion, leaderAndIsrRequests.head.version)
```
and then checks the expected version against the one the test used to build the request. What is the assertion here? Do we expect the `build()` to fail if the test gives it the new version but the production code used the old version?

ijuma commented:
How is the change compatible? If you added any fields, it would not be compatible.

cmccabe commented:
LeaderAndIsrRequest version 3 is not part of Kafka 2.2, so this check is not correct.  It will be part of Kafka 2.4, so you need to check for that.

I think we need some kind of integration test here to verify that this is working as we expect.  Something like doing a reassignment and verifying that the other brokers get the AR and RR.

stanislavkozlovski commented:
Sorry, I thought we didn't need to bump IBP but that doesn't make sense. I confused it with previous discussions around the KIP which discussed not needing to check IBP before writing the new fields to the znode - that's why I didn't think we need to bump it in the original PR

stanislavkozlovski committed:
Add new IBP version (2.4) and add test for LAIR request reassignment fields

stanislavkozlovski commented:
Not ready for a review yet, I'm syncing offline to understand whether I need to change ControllerChannelManagerTest - I still don't completely understand what it does as is the code right now

hachikuji committed:
MINOR: Fix broken version assertions in `ControllerChannelManagerTest`

ijuma commented:
Also see https://github.com/apache/kafka/pull/7359

hachikuji reviewed (commented):
I guess until KIP-352 is implemented, we have no functional dependence on the adding/removing sets that come from LeaderAndIsr?

stanislavkozlovski commented:
@hachikuji yes - that's why no integration tests broke

stanislavkozlovski commented:
I cherry-picked the commit from #7359 after syncing with @hachikuji. I verified the now-fixed test fails with the old code:
```
java.lang.AssertionError: IBP 2.4-IV0 should use version 3 
```
@cmccabe we don't necessarily need an integration/other kind of test to verify this behavior now as we've fixed ControllerChannelManagerTest, right?

Call for reviews - @hachikuji @cmccabe @ijuma

stanislavkozlovski commented:
JDK 8 passed.

JDK 11 / 2.12 - expired build, can't see it

JDK11 / 2.13 timed out after 4hr 30m, with 1 failure:
```
22:20:06 Build timed out (after 270 minutes). Marking the build as aborted.
```

retest this please

hachikuji reviewed (approved):
LGTM
